### PR TITLE
Update function formatting tactic heuristics

### PR DESCRIPTION
**Pull request by marcusklaas**
Close https://github.com/rust-lang-nursery/rustfmt/issues/692.
Close https://github.com/rust-lang-nursery/rustfmt/issues/371.
Close https://github.com/rust-lang-nursery/rustfmt/issues/501.

cc https://github.com/rust-lang-nursery/rustfmt/issues/450.

### CODE REVIEW

File: src/lists.rs [hunk 1]
```diff
@@ -167,21 +169,35 @@ pub fn definitive_tactic<'t, I, T>(items: I,
...
-    let (sep_count, total_width) = calculate_width(items.clone());
     let sep_len = ", ".len(); // FIXME: make more generic?
     let total_sep_len = sep_len * sep_count.checked_sub(1).unwrap_or(0);
     let real_total = total_width + total_sep_len;
 
-    if real_total <= limit && !pre_line_comments &&
+    let horizont_check = match tactic {
+        ListTactic::LimitedHorizontalVertical(limit, l) => {
+            item_width_vec.iter().skip(1).all(|&w| w <= l) ||
+            item_width_vec.iter().rev().skip(1).all(|&w| w <= l) || real_total <= limit
```
Comment (nrc) [inline src/lists.rs]: what is going here? Why the `rev()`?

Comment (marcusklaas) [inline src/lists.rs]: That takes basically all but the last elements and checks that they are not longer than a given limit.

Comment (marcusklaas) [inline src/lists.rs]: Correction: all but the first!

**Comment by nrc** [comment]
This is an improvement, but I worry that we may introduce some worse cases where there are many short-ish args after a long one. A simple tweak would be to only apply this if there are two args (i.e., a long one and a short one), but if there are >= 3, then we ignore the lower limit. We could also do a little more sophisticated maths, e.g., have a soft and hard upper limit (the current heuristic is the soft upper) and we use the hard upper limit if the other args are short.

File: src/lists.rs (hunk 1 above)
Comment (nrc) [inline src/lists.rs]: Isn't that what the first clause does? Then what does the second clause do?

Comment (marcusklaas) [inline src/lists.rs]: The first checks all but the first, the second checks all but the last.

Comment (nrc) [inline src/lists.rs]: Why check all but the last? I guess on reflection, why check all but the first too?

**Comment by nrc** [comment]
ping @marcusklaas do you have plans for this PR?

**Comment by marcusklaas** [comment]
Not really - I'll close this for now.

**Pull request closed without merging**